### PR TITLE
MAGENTO2-151

### PR DESCRIPTION
**Pull request by Skistims**
Support of cart rule

### CODE REVIEW

File: src/Helper/Quote.php [hunk 1]
```diff
@@ -109,16 +109,42 @@ public function __construct(
     public function getValidatedCoupons()
     {
         $coupons = [];
+        $returnInvalidCartRuleCoupon = false;
+        $amount = $this->quote->getSubtotal() - $this->quote->getSubtotalWithDiscount();
+        $couponsIncludeTax = $this->taxHelper->couponInclTax();
         foreach ($this->sgBase->getExternalCoupons() as $coupon) {
+            if ($coupon->getCode() == '1'){
```
Comment (augsteyer) [inline src/Helper/Quote.php]: the `1` needs to become a constant

File: src/Helper/Quote.php [hunk 2]
```diff
@@ -109,16 +109,42 @@ public function __construct(
     public function getValidatedCoupons()
     {
         $coupons = [];
+        $returnInvalidCartRuleCoupon = false;
```
Comment (augsteyer) [inline src/Helper/Quote.php]: I think the `=` signs need to align, I cannot remember for sure, but I think this project styles are like that

File: src/Helper/Quote.php [hunk 3]
```diff
@@ -109,16 +109,42 @@ public function __construct(
...
+        $amount = $this->quote->getSubtotal() - $this->quote->getSubtotalWithDiscount();
```
Comment (augsteyer) [inline src/Helper/Quote.php]: maybe needs to be renamed this `amount` variable, amount of what? Maybe someone who does not know mage will not understand that this is discount. 
P.S. yeah, I know this is my sloppy code copied over

File: src/Helper/Quote.php [hunk 4]
```diff
@@ -109,16 +109,42 @@ public function __construct(
...
             }
             $coupons[] = $this->couponHelper->dataToEntity($coupon->toArray());
         }
+        if (empty($coupons) && !empty($amount)) {
+            $coupon = new \ShopgateExternalCoupon();
+            $coupon->setCode('1');
+            $coupon->setIsValid(true);
+            $coupon->setCurrency($this->quote->getStoreCurrencyCode());
+            $coupon->setIsFreeShipping((bool) $this->quote->getShippingAddress()->getFreeShipping());
+            $couponsIncludeTax ? $coupon->setAmountGross($amount) : $coupon->setAmountNet($amount);
```
Comment (augsteyer) [inline src/Helper/Quote.php]: there seems to be a lot of copy pasting here, can you refactor this logic to a function by chance?

**Review by augsteyer — CHANGES_REQUESTED** [review]

File: src/Helper/Quote.php (hunk 1 above)
Comment (augsteyer) [reply src/Helper/Quote.php]: also needs to be a strict comparison `===`

**Comment by augsteyer** [comment]
in addition, please fix the issue with failing tests. The `src/Helper/Product/Retriever.php` is using the old `array` syntax. This was introduced in the earlier commit.

**Comment by augsteyer** [comment]
There is also seem to be a lack of option to disable cart rules. Currently they do not work, so this will enable them for all customers. Not sure if that's the desired behavior.

File: src/Helper/Quote.php (hunk 2 above)
Comment (Skistims) [reply src/Helper/Quote.php]: done

File: src/Helper/Quote.php (hunk 3 above)
Comment (Skistims) [reply src/Helper/Quote.php]: k, $discountAmount it is

File: src/Helper/Quote.php (hunk 1 above)
Comment (Skistims) [reply src/Helper/Quote.php]: done

File: src/Helper/Quote.php (hunk 4 above)
Comment (Skistims) [reply src/Helper/Quote.php]: how is it now?

Comment (augsteyer) [reply src/Helper/Quote.php]: are you saying you are done and I can review?

Comment (Skistims) [reply src/Helper/Quote.php]: Sorry for the confusion. No.
I will leave a comment that I am done with the feedback when I am ready.

**Comment by Skistims** [comment]
fixed the test issue.
Please review again @augsteyer

File: src/Helper/Quote.php [hunk 5]
```diff
@@ -108,17 +112,51 @@ public function __construct(
...
+                'name'             => __(self::CART_RULE_COUPON_NAME),
+                'is_valid'         => true,
+                'currency'         => $quoteCurrency,
+                'is_free_shipping' => (bool)$this->quote->getShippingAddress()->getFreeShipping(),
+                'amount_gross'     => $couponsIncludeTax ? $discountAmount : null,
+                'amount_net'       => $couponsIncludeTax ? null : $discountAmount,
+
+            ];
+            $coupons[]                   = $this->couponHelper->dataToEntity($couponData);
+            $returnInvalidCartRuleCoupon = false;
```
Comment (augsteyer) [inline src/Helper/Quote.php]: just one last thing, can you make this variable name shorter? It is just way too long. Smth like `$isCouponValid`

**Review by augsteyer — APPROVED** [review]

**Pull request merged**